redisson/src/main/java/org/redisson/RedissonExecutorService.java

```diff
@@ -854,7 +815,7 @@ private void storeReference(RExecutorFuture<?> future, RequestId requestId) {
             throw new IllegalArgumentException("Wrong cron expression! Unable to calculate start date");
         }
         long startTime = startDate.getTime();
-        RemotePromise<Void> result = (RemotePromise<Void>) asyncScheduledServiceAtFixed.schedule(task.getClass().getName(), classBody, state, startTime, cronSchedule.getExpression().getCronExpression(), executorId, null);
+        RemotePromise<Void> result = (RemotePromise<Void>) asyncScheduledServiceAtFixed.schedule(task.getClass().getName(), classBody, state, startTime, cronSchedule.getExpression(), executorId, null);
```

Comment:
It's a bad idea to pass whole CronExpression object. Pleasae create separate object for it

Reply:
Inside TaskRunnerService at [line-123](https://github.com/redisson/redisson/blob/46e8e8846a260999b6706e331b40943de0ffb813/redisson/src/main/java/org/redisson/executor/TasksRunnerService.java#L123) we are actually creating CronExpression object every time which I feel is unnecessary and can be avoided. And as we need the full CronExpression object there so wouldn't it be good to pass it from outside instead of constructing it every time there.
 
Let me know your views on this, I will make the changes accordingly.

Reply:
@arpit728 Serialization of CronExpression object involves some customization like you did above to make it compatibility with Jackson. I want to avoid it. Please pass it as String as well as TimeZone.

Reply:
@mrniko
I have added timezone as a separate parameter, but getting following error.
```
[redisson-5-6] ERROR org.redisson.RedissonRemoteService - Can't execute: RemoteServiceRequest [requestId=010cee9ccf41288ebd1431695ed4265ff2, methodName=schedule, signatures=[[java.lang.String, [B, [B, long, java.lang.String, java.lang.String, java.lang.String, java.lang.String]], args=[com.webengage.DemoJob, [B@32328c45, [B@865bbae, 1532310600000, 0 20,22 7 23 JUL ? 2018, Asia/Kolkata, 010cee9ccf41288ebd1431695ed4265ff2, null], options=RemoteInvocationOptions[ackTimeoutInMillis=null, executionTimeoutInMillis=null], date=1532310552827]
java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.redisson.RedissonRemoteService.invokeMethod(RedissonRemoteService.java:347)
	at org.redisson.RedissonRemoteService.access$400(RedissonRemoteService.java:66)
	at org.redisson.RedissonRemoteService$2.run(RedissonRemoteService.java:312)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run$$$capture(FutureTask.java:266)
	at java.util.concurrent.FutureTask.run(FutureTask.java)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at io.netty.buffer.ByteBufUtil.decodeHexDump(ByteBufUtil.java:141)
	at org.redisson.remote.RequestId.<init>(RequestId.java:32)
	at org.redisson.executor.TasksRunnerService.asyncScheduledServiceAtFixed(TasksRunnerService.java:156)
	at org.redisson.executor.TasksRunnerService.schedule(TasksRunnerService.java:128)
	... 14 more
[redisson-5-7] INFO com.webengage.DemoJob - Logging working fine
Job executed successfullyJobTemplate(jobType=HTTP, targetUrl=null, headers={}, queryString=null, httpMethod=null, postData=null, callbackUrl=https://sample-url)
[redisson-netty-4-8] WARN io.netty.channel.AbstractChannelHandlerContext - Failed to mark a promise as failure because it has failed already: DefaultChannelPromise@2cb69f94(failure: io.netty.handler.codec.EncoderException: java.lang.NullPointerException), unnotified cause: io.netty.handler.codec.EncoderException: java.lang.NullPointerException
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:125)
	at org.redisson.client.handler.CommandEncoder.write(CommandEncoder.java:73)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:120)
	at org.redisson.client.handler.CommandBatchEncoder.write(CommandBatchEncoder.java:45)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
	at io.netty.channel.ChannelDuplexHandler.write(ChannelDuplexHandler.java:106)
	at org.redisson.client.handler.CommandsQueue.write(CommandsQueue.java:92)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:801)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:814)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:794)
	at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1066)
	at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:305)
	at org.redisson.client.handler.CommandsQueue.sendData(CommandsQueue.java:118)
	at org.redisson.client.handler.CommandsQueue.write(CommandsQueue.java:95)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.access$1900(AbstractChannelHandlerContext.java:38)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.write(AbstractChannelHandlerContext.java:1081)
	at io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.write(AbstractChannelHandlerContext.java:1128)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:1070)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:465)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.redisson.client.handler.CommandEncoder.encode(CommandEncoder.java:124)
	at org.redisson.client.handler.CommandEncoder.encode(CommandEncoder.java:97)
	at org.redisson.client.handler.CommandEncoder.encode(CommandEncoder.java:53)
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:107)
	... 34 more

io.netty.handler.codec.EncoderException: java.lang.NullPointerException
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:125)
	at org.redisson.client.handler.CommandEncoder.write(CommandEncoder.java:73)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:120)
	at org.redisson.client.handler.CommandBatchEncoder.write(CommandBatchEncoder.java:45)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:816)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:723)
	at io.netty.channel.ChannelDuplexHandler.write(ChannelDuplexHandler.java:106)
	at org.redisson.client.handler.CommandsQueue.write(CommandsQueue.java:92)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWriteAndFlush(AbstractChannelHandlerContext.java:801)
	at io.netty.channel.AbstractChannelHandlerContext.write(AbstractChannelHandlerContext.java:814)
	at io.netty.channel.AbstractChannelHandlerContext.writeAndFlush(AbstractChannelHandlerContext.java:794)
	at io.netty.channel.DefaultChannelPipeline.writeAndFlush(DefaultChannelPipeline.java:1066)
	at io.netty.channel.AbstractChannel.writeAndFlush(AbstractChannel.java:305)
	at org.redisson.client.handler.CommandsQueue.sendData(CommandsQueue.java:118)
	at org.redisson.client.handler.CommandsQueue.write(CommandsQueue.java:95)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite0(AbstractChannelHandlerContext.java:738)
	at io.netty.channel.AbstractChannelHandlerContext.invokeWrite(AbstractChannelHandlerContext.java:730)
	at io.netty.channel.AbstractChannelHandlerContext.access$1900(AbstractChannelHandlerContext.java:38)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.write(AbstractChannelHandlerContext.java:1081)
	at io.netty.channel.AbstractChannelHandlerContext$WriteAndFlushTask.write(AbstractChannelHandlerContext.java:1128)
	at io.netty.channel.AbstractChannelHandlerContext$AbstractWriteTask.run(AbstractChannelHandlerContext.java:1070)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute$$$capture(AbstractEventExecutor.java:163)
	at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java)
	at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
	at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:465)
	at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:884)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.lang.NullPointerException
	at org.redisson.client.handler.CommandEncoder.encode(CommandEncoder.java:124)
	at org.redisson.client.handler.CommandEncoder.encode(CommandEncoder.java:97)
	at org.redisson.client.handler.CommandEncoder.encode(CommandEncoder.java:53)
	at io.netty.handler.codec.MessageToByteEncoder.write(MessageToByteEncoder.java:107)
	... 34 more
```

Reply:
@arpit728 

Seems you found a bug, could you please give me a test to reproduce it?

Reply:
@mrniko
To reproduce just schedule a cron job in the way described below.
```java
public static void main(String args[]){
    CronSchedule cronSchedule = CronSchedule.of("0 0,1,2,3 11 19 AUG ? 2018");
    cronSchedule.getExpression().setTimeZone(TimeZone.getTimeZone("Asia/Kolkata"));
    RExecutorFuture r = (RExecutorFuture) ((RScheduledExecutorService) e).schedule(new DemoJob(), cronSchedule);
    System.out.println(r.getTaskId());
}
public class DemoJob implements Runnable{
    public DemoJob() { }

    @Override
    public void run() {
        System.out.println("Job executed successfully.");
    }
}
```

Reply:
@arpit728  I'm getting "Wrong cron expression! Unable to calculate start date" error with 3.7.5 version

Reply:
```CronSchedule cronSchedule = CronSchedule.of("0 0,1,2,3 11 19 AUG ? 2018");```

Did you replaced the cronExpression in the above line with the valid cronExpression, I used this expression at the time of testing, now this has become invalid as no executions for it is going to come in future.

Reply:
@arpit728 

Here is my test: 
    @Test
    public void test1() {
        CronSchedule cronSchedule = CronSchedule.of("0 0,1,2,3 11 22 AUG ? 2018");
        cronSchedule.getExpression().setTimeZone(TimeZone.getTimeZone("Asia/Kolkata"));
        RExecutorFuture r = redisson.getExecutorService("test2").schedule(new RunnableTask(), cronSchedule);
        System.out.println(r.getTaskId());
    }

I added it to `RedissonScheduledExecutorServiceTest` and couldn't to reproduce it.